File: Day 19/day19.pl
use v5.30;
use Data::Dumper;
use File::Slurp;
use String::Scanf;

my $res = 0;
for my $line (split '\n', read_file('day19.txt')) {
  my $idx;
  my @costs = ([0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]);
  ($idx, $costs[0][0], $costs[1][0], $costs[2][0], $costs[2][1], $costs[3][0], $costs[3][2]) 
    = sscanf("Blueprint %d: Each ore robot costs %d ore. Each clay robot costs %d ore. Each obsidian robot costs %d ore and %d clay. Each geode robot costs %d ore and %d obsidian.", $line);
  
  my $max_geode = 0;
  my $p3 = 1;
  for my $len (1 .. 17) {
    $p3 *= 3;
    for my $sm (0 .. $p3 - 1) {
      my @robot_order;

      my $ok = 1;
      my @fr = (0, 0, 0, 0);
      my $csm = $sm;
      for my $i (0 .. $len - 1) {
        my $crobot = $csm % 3;
        push @robot_order, $crobot;
        ++$fr[$crobot];
        $ok = 0 if $fr[1] + $fr[2] > 0 && !$fr[0] || !$fr[1] && $fr[2];
        $ok = 0 if $fr[1] > $costs[2][1] || $fr[2] > $costs[3][2] || ($fr[0] > $costs[0][0] && $fr[0] > $costs[2][0] && $fr[0] > $costs[3][0]);
        last unless $ok;
        $csm /= 3;
      }

      next unless $ok;
      # say join ',', @robot_order;
      # my @robot_order = (1, 1, 1, 2, 1, 2);
      # say "GOOD ARRAY" if ($robot_order[0] == 1 && $robot_order[1] == 1 && $robot_order[2] == 1 && $robot_order[3] == 2 && $robot_order[4] == 1 && $robot_order[5] == 2 && @robot_order == 6);
      my @resources = (0, 0, 0, 0);
      my @robots = @resources;
      $robots[0] = 1;
      for my $minute (1 .. 24) {
        my @new_resources = @resources;
        $new_resources[$_] += $robots[$_] for (0 .. 3);
        if (@robot_order && $resources[0] >= $costs[$robot_order[0]][0] && $resources[1] >= $costs[$robot_order[0]][1] && $resources[2] >= $costs[$robot_order[0]][2] && $resources[3] >= $costs[$robot_order[0]][3]) {
          $new_resources[0] -= $costs[$robot_order[0]][0];
          $new_resources[1] -= $costs[$robot_order[0]][1];
          $new_resources[2] -= $costs[$robot_order[0]][2];
          $new_resources[3] -= $costs[$robot_order[0]][3];
          ++$robots[$robot_order[0]];
          shift @robot_order;
        }

        if (scalar @robot_order == 0 && $resources[0] >= $costs[3][0] && $resources[1] >= $costs[3][1] && $resources[2] >= $costs[3][2] && $resources[3] >= $costs[3][3]) {
          $new_resources[0] -= $costs[3][0];
          $new_resources[1] -= $costs[3][1];
          $new_resources[2] -= $costs[3][2];
          $new_resources[3] -= $costs[3][3];
          ++$robots[3];
        }
        @resources = @new_resources;
        $max_geode = $resources[3] if $max_geode < $resources[3];

        # if ($robot_order[0] == 1 && $robot_order[1] == 1 && $robot_order[2] == 1 && $robot_order[3] == 2 && $robot_order[4] == 1 && $robot_order[5] == 2) {
        #   say "minute: $minute";
        #   say join "|", @robots;
        #   say join "#", @resources;
        #   say "$resources[0] >= $costs[3][0] && $resources[1] >= $costs[3][1] && $resources[2] >= $costs[3][2] && $resources[3] >= $costs[3][3]";
        #   exit if $minute == 24;
        # }
      }
    }
  }
  say $max_geode;
  $res += $idx * $max_geode;
}

say $res;

File: Day 19/day19.cpp
#include <iostream>
#include <fstream>
#include <cstdio>
using namespace std;

int costs[4][4];
int robots[4], resources[4];
int max_geodes;
int robot_order[40], n_robots;
long long prints = 0;

void generate(int minute, int last_robot = 1) {
  // ++prints;
  // if (prints % 10000000 == 0) {
  //   cout << minute << endl;
  //   for (int i = 1; i <= n_robots; ++i)
  //     cout << robot_order[i] << " ";
  //   cout << endl;
  // }
  max_geodes = max(max_geodes, resources[3]);
  if (minute == 25) {
    return;
  }
  int c_robots[4]; 
  for (int j = 0; j < 4; ++j) {
    c_robots[j] = robots[j];
  }
  int generated = 0;
  int start = max(0, last_robot - 1), end = min(last_robot + 1, 3);
  if (n_robots > 2 && robot_order[n_robots] == robot_order[n_robots - 1] && robot_order[n_robots] == robot_order[n_robots - 2]) {
    start = last_robot;
  }
  for (int next_robot = start; next_robot <= end; ++next_robot) {
    int ok = 1;
    for (int i = 0; i < next_robot; ++i) {
      if (!robots[i]) ok = 0;
    }
    if (ok && (robots[next_robot] < max(max(costs[0][next_robot], costs[1][next_robot]), max(costs[2][next_robot], costs[3][next_robot])) || next_robot == 3)
      && resources[0] >= costs[next_robot][0] 
      && resources[1] >= costs[next_robot][1] 
      && resources[2] >= costs[next_robot][2] 
      && resources[3] >= costs[next_robot][3]) {
        ++generated;
        resources[0] -= costs[next_robot][0];
        resources[1] -= costs[next_robot][1];
        resources[2] -= costs[next_robot][2];
        resources[3] -= costs[next_robot][3];
        ++robots[next_robot];
        robot_order[++n_robots] = next_robot;
        for (int j = 0; j < 4; ++j) {
          resources[j] += c_robots[j];
        }
        generate(minute + 1, next_robot);
        --n_robots;
        --robots[next_robot];
        resources[0] += costs[next_robot][0];
        resources[1] += costs[next_robot][1];
        resources[2] += costs[next_robot][2];
        resources[3] += costs[next_robot][3];
        for (int j = 0; j < 4; ++j) {
          resources[j] -= c_robots[j];
        }
      }
  }
  if (generated < end - start + 1) {
    for (int j = 0; j < 4; ++j) {
      resources[j] += c_robots[j];
    }
    generate(minute + 1, last_robot);
    for (int j = 0; j < 4; ++j) {
      resources[j] -= c_robots[j];
    }
  }
}

int main() {
  int res = 0;
  char blueprint_string[300];
  ifstream fin("day19.txt");
  while(fin.getline(blueprint_string, 300)) {
    int idx = 0;
    max_geodes = 0;
    sscanf(blueprint_string, "Blueprint %d: Each ore robot costs %d ore. Each clay robot costs %d ore. Each obsidian robot costs %d ore and %d clay. Each geode robot costs %d ore and %d obsidian.", 
      &idx, &costs[0][0], &costs[1][0], &costs[2][0], &costs[2][1], &costs[3][0], &costs[3][2]);
    robots[0] = robots[1] = robots[2] = robots[3] = 0;
    resources[0] = resources[1] = resources[2] = resources[3] = 0;
    robots[0] = 1;
    n_robots = 0;
    generate(1);
    cout << "Geodes: " << max_geodes << "\n"; cout.flush();
    res += idx * max_geodes;
  }
  cout << res;
}